cli_tools/common/utils/daisyutils/daisy_utils.go

```diff
@@ -355,19 +355,27 @@ func PostProcessDErrorForNetworkFlag(action string, err error, network string, w
 	}
 }
 
-// RunWorkflowWithCancelSignal runs Daisy workflow with accepting Ctrl-C signal
-func RunWorkflowWithCancelSignal(ctx context.Context, w *daisy.Workflow) error {
+// RunWorkflowWithCancelSignal runs a Daisy workflow, and allows for cancellation from two sources:
+//   1. The user types Ctrl-C on their keyboard.
+//   2. The caller sends a cancellation reason on the cancel channel (or closes it).
+func RunWorkflowWithCancelSignal(w *daisy.Workflow, cancel <-chan string) error {
 	c := make(chan os.Signal, 1)
 	signal.Notify(c, os.Interrupt)
 	go func(w *daisy.Workflow) {
 		select {
+		case reason := <-cancel:
+			if reason != "" {
+				w.CancelWithReason(reason)
+			}
+			break
 		case <-c:
 			w.LogWorkflowInfo("\nCtrl-C caught, sending cancel signal to %q...\n", w.Name)
 			w.CancelWorkflow()
+			break
 		case <-w.Cancel:
 		}
 	}(w)
-	return w.Run(ctx)
+	return w.Run(context.Background())
```

Comment:
No need to reuse the context here?

Reply:
Good question -- in short, the context that's passed here shouldn't be cancelled. If cancellation occurs, then all of the clients within daisy die, and cleanup can't occur. I added a comment to explain why we're adding a new context.